### PR TITLE
[Enhancement] array_agg supports distinct

### PR DESCRIPTION
**Pull request by fzhedu**
array_agg support distinct
1. for complex types and more than one inputs, rewrite it to group by
2. just one input with non-nest types, use array_agg_distinct

```
mysql> select array_agg(name order by 1), array_agg(distinct name order by 1, score) from ss group by id order by 1;
+------------------------------------------------------+-------------------------------------------------------+
| array_agg(name ORDER BY name ASC)                    | array_agg(DISTINCT name ORDER BY name ASC, score ASC) |
+------------------------------------------------------+-------------------------------------------------------+
| [null]                                               | [null]                                                |
| [null,"Ti"]                                          | [null,"Ti"]                                           |
| ["May","Ti","欧阳诸葛方程"]                          | ["May","Ti","欧阳诸葛方程"]                           |
| ["Tom","Tom"]                                        | ["Tom"]                                               |
| ["Tom","Tom","王武程咬金"]                           | ["Tom","王武程咬金"]                                  |
| ["张三掩耳盗铃","张三此地无银三百两"]                | ["张三掩耳盗铃","张三此地无银三百两"]                 |
| ["李四大闹天空"]                                     | ["李四大闹天空"]                                      |
+------------------------------------------------------+-------------------------------------------------------+
```
7 rows in set (0.09 sec)

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [x] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 3 / 13 (23.08%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exprs/agg/array_agg.h|3|13|23.08%|[254, 255, 256, 257, 258, 259, 260, 261, 265, 266]|

File: be/src/exprs/agg/array_agg.h [hunk 1]
```diff
@@ -249,6 +248,25 @@ class ArrayAggAggregateFunctionV2
             materialize_column_by_permutation(tmp.get(), {(*state_impl.data_columns)[0]}, perm);
             res = ColumnPtr(std::move(tmp));
         }
+        // further remove duplicated values, pick the last unique one to identify the last sep and don't output it.
+        // TODO(fzh) optimize it later
+        if (ctx->get_is_distinct()) {
```
Comment (satanson) [inline be/src/exprs/agg/array_agg.h:253]: use is_distinct instead of get_is_distinct

File: be/src/exprs/agg/array_agg.h [hunk 2]
```diff
@@ -249,6 +248,25 @@ class ArrayAggAggregateFunctionV2
...
+            bool is_duplicated = false;
+            Filter filter(elem_size, 1);
+            for (auto row_id = 0; row_id < elem_size; row_id++) {
```
Comment (satanson) [inline be/src/exprs/agg/array_agg.h:256]: use a sort-based deduplication or hash-based(flat_hash_map) deduplication algorithm instead of nested loop

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 16 / 23 (69.57%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/SplitAggregateRule.java|0|1|00.00%|[320]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|1|3|33.33%|[1904, 1905]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|11|15|73.33%|[1244, 1245, 1252, 1253]|
|:large_blue_circle:|com/starrocks/sql/analyzer/SelectAnalyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AstToStringBuilder.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctByCTERule.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 3 / 13 (23.08%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exprs/agg/array_agg.h|3|13|23.08%|[254, 255, 256, 257, 258, 259, 260, 261, 265, 266]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctByCTERule.java [hunk 3]
```diff
@@ -280,6 +280,7 @@ private LinkedList<OptExpression> buildDistinctAggCTEConsume(LogicalAggregationO
             CallOperator aggCallOperator = aggregate.getAggregations().get(distinctAggRef);
             if (aggCallOperator.getFnName().equalsIgnoreCase(FunctionSet.COUNT) ||
                     aggCallOperator.getFnName().equalsIgnoreCase(FunctionSet.SUM) ||
+                    aggCallOperator.getFnName().equalsIgnoreCase(FunctionSet.ARRAY_AGG) ||
                     aggCallOperator.getFnName().equalsIgnoreCase(FunctionSet.GROUP_CONCAT)) {
                 allCteConsumes.offer(buildCountSumDistinctCTEConsume(distinctAggRef, aggCallOperator,
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctByCTERule.java:285]: We'd better rename this method.

**Review by packy92 — DISMISSED** [review]

**Review by kangkaisen — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 16 / 23 (69.57%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/SplitAggregateRule.java|0|1|00.00%|[320]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|1|3|33.33%|[1904, 1905]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|11|15|73.33%|[1244, 1245, 1252, 1253]|
|:large_blue_circle:|com/starrocks/sql/analyzer/SelectAnalyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AstToStringBuilder.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctByCTERule.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 3 / 13 (23.08%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exprs/agg/array_agg.h|3|13|23.08%|[254, 255, 256, 257, 258, 259, 260, 261, 265, 266]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 16 / 23 (69.57%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/SplitAggregateRule.java|0|1|00.00%|[320]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|1|3|33.33%|[1904, 1905]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|11|15|73.33%|[1244, 1245, 1252, 1253]|
|:large_blue_circle:|com/starrocks/sql/analyzer/SelectAnalyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AstToStringBuilder.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctByCTERule.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 3 / 13 (23.08%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exprs/agg/array_agg.h|3|13|23.08%|[254, 255, 256, 257, 258, 259, 260, 261, 265, 266]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 3 / 13 (23.08%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exprs/agg/array_agg.h|3|13|23.08%|[254, 255, 256, 257, 258, 259, 260, 261, 265, 266]|

File: be/src/exprs/agg/array_agg.h (hunk 2 above)
Comment (fzhedu) [reply be/src/exprs/agg/array_agg.h:318]: general imp for all types.

**Review by satanson — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 3 / 13 (23.08%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exprs/agg/array_agg.h|3|13|23.08%|[254, 255, 256, 257, 258, 259, 260, 261, 265, 266]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 3 / 13 (23.08%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exprs/agg/array_agg.h|3|13|23.08%|[254, 255, 256, 257, 258, 259, 260, 261, 265, 266]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 40 / 66 (60.61%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/aggregator.cpp|0|3|00.00%|[400, 402, 403]|
|:large_blue_circle:|src/exprs/agg/array_agg.h|31|54|57.41%|[67, 69, 70, 73, 129, 130, 156, 314, 315, 316, 317, 318, 319, 320, 321, 322, 324, 325, 326, 327, 328, 332, 333]|
|:large_blue_circle:|src/exprs/agg/factory/aggregate_resolver_avg.cpp|9|9|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 46 / 60 (76.67%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctRule.java|1|7|14.29%|[88, 188, 189, 190, 192, 193]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/SplitAggregateRule.java|1|4|25.00%|[320, 321, 322]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctByCTERule.java|1|2|50.00%|[290]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|11|15|73.33%|[1244, 1245, 1252, 1253]|
|:large_blue_circle:|com/starrocks/sql/analyzer/SelectAnalyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|19|19|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/DecimalV3FunctionAnalyzer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AstToStringBuilder.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/CostModel.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 40 / 66 (60.61%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/aggregator.cpp|0|3|00.00%|[400, 402, 403]|
|:large_blue_circle:|src/exprs/agg/array_agg.h|31|54|57.41%|[67, 69, 70, 73, 129, 130, 156, 314, 315, 316, 317, 318, 319, 320, 321, 322, 324, 325, 326, 327, 328, 332, 333]|
|:large_blue_circle:|src/exprs/agg/factory/aggregate_resolver_avg.cpp|9|9|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 40 / 66 (60.61%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/aggregator.cpp|0|3|00.00%|[400, 402, 403]|
|:large_blue_circle:|src/exprs/agg/array_agg.h|31|54|57.41%|[67, 69, 70, 73, 129, 130, 156, 314, 315, 316, 317, 318, 319, 320, 321, 322, 324, 325, 326, 327, 328, 332, 333]|
|:large_blue_circle:|src/exprs/agg/factory/aggregate_resolver_avg.cpp|9|9|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 57 / 62 (91.94%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctByCTERule.java|1|2|50.00%|[290]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|11|15|73.33%|[1260, 1261, 1268, 1269]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/SplitAggregateRule.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/SelectAnalyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|19|19|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/DecimalV3FunctionAnalyzer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AstToStringBuilder.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctRule.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/CostModel.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 40 / 66 (60.61%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/aggregator.cpp|0|3|00.00%|[400, 402, 403]|
|:large_blue_circle:|src/exprs/agg/array_agg.h|31|54|57.41%|[67, 69, 70, 73, 129, 130, 156, 314, 315, 316, 317, 318, 319, 320, 321, 322, 324, 325, 326, 327, 328, 332, 333]|
|:large_blue_circle:|src/exprs/agg/factory/aggregate_resolver_avg.cpp|9|9|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 57 / 62 (91.94%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctByCTERule.java|1|2|50.00%|[290]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|11|15|73.33%|[1269, 1270, 1277, 1278]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/SplitAggregateRule.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/SelectAnalyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|19|19|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/DecimalV3FunctionAnalyzer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AstToStringBuilder.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctRule.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/CostModel.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 40 / 66 (60.61%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/aggregator.cpp|0|3|00.00%|[400, 402, 403]|
|:large_blue_circle:|src/exprs/agg/array_agg.h|31|54|57.41%|[67, 69, 70, 73, 129, 130, 156, 314, 315, 316, 317, 318, 319, 320, 321, 322, 324, 325, 326, 327, 328, 332, 333]|
|:large_blue_circle:|src/exprs/agg/factory/aggregate_resolver_avg.cpp|9|9|100.00%|[]|

**Review by satanson — APPROVED** [review]

**Review by packy92 — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Comment by fzhedu** [comment]
https://github.com/Mergifyio backport branch-3.0

**Pull request merged**